File: PyProject/custom/quick_sort.py
def partition(arr, low, high):
    i = low - 1

    pivot = arr[high]
    for j in range(low, high):
        if arr[j] < pivot:
            arr[i + 1], arr[j] = arr[j], arr[i + 1]
            i += 1

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quick_sort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)


if __name__ == '__main__':
    arr = [1, 5, 2, 8, 0, 6, 3]
    quick_sort(arr, 0, len(arr) - 1)
    print(arr)

File: GoProject/leetcode/sort_dir/quick_sort.go
package main

import "fmt"

func partition(arr []int, start int, end int) int {
	i := start - 1
	pivot := arr[end]

	for j := start; j < end; j++ {
		if arr[j] <= pivot {
			i++
			arr[j], arr[i] = arr[i], arr[j]
		}
	}
	arr[end], arr[i+1] = arr[i+1], arr[end]
	return i + 1

}

func quickSort(arr []int, start int, end int) {
	if start < end {
		pivot := partition(arr, start, end)

		quickSort(arr, start, pivot-1)
		quickSort(arr, pivot+1, end)
	}
}
func main() {
	arr := []int{10, 7, 8, 9, 1, 5}
	quickSort(arr, 0, len(arr)-1)
	fmt.Println(arr)
}
